src/main/kotlin/nl/nuts/discovery/service/CertificatesApiServiceImpl.kt

```diff
@@ -19,23 +19,280 @@
+
+    @Autowired
+    lateinit var nutsCertificateRequestRepository: NutsCertificateRequestRepository
+
+    @Autowired
+    lateinit var certificateRepository: CertificateRepository
+
+    @Autowired
+    lateinit var cASerialRepository: CustomCASerialRepository
+
     override fun listCertificates(otherName: String): List<CertificateWithChain> {
-        TODO("Not yet implemented")
+        return certificateRepository.findByOid(otherName).map {
+            CertificateWithChain(it.toPem(), splitChain(it.chain))
+        }
     }
 
     override fun listRequests(otherName: String): List<CertificateSigningRequest> {
-        TODO("Not yet implemented")
+        return nutsCertificateRequestRepository.findByOid(otherName).map {
+            entityToApiModel(it)
+        }
     }
 
     override fun submit(body: String): CertificateSigningRequest {
-        TODO("Not yet implemented")
+        val nutsCertificateRequest = NutsCertificateRequest.fromPEM(body)
+
+        logger.info("Received request for: ${nutsCertificateRequest.name}, oid: ${nutsCertificateRequest.oid}")
+
+        nutsCertificateRequestRepository.save(nutsCertificateRequest)
+        if (nutsDiscoveryProperties.autoAck) {
+             sign(nutsCertificateRequest)
+        }
+
+        return entityToApiModel(nutsCertificateRequest)
+    }
+
+    private fun entityToApiModel(nutsCertificateRequest: NutsCertificateRequest) : CertificateSigningRequest {
+        return CertificateSigningRequest(
+            subject = nutsCertificateRequest.name!!,
+            pem = nutsCertificateRequest.pem!!,
+            submittedAt = nutsCertificateRequest.submittedAt.toString()
+        )
+    }
+
+    override fun sign(request: NutsCertificateRequest) : X509Certificate {
+        // some checks
+        // todo additional checks?
+        if (request.oid == null) {
+            throw java.lang.IllegalArgumentException("missing oid")
+        }
+
+        // convert
+        val pkcs10 = JcaPKCS10CertificationRequest(request.toPKCS10())
+
+        // x509 builder
+        val certificateBuilder = certificateBuilderWithDefaults(pkcs10)
+        addCustomExtensions(certificateBuilder, request.oid!!, pkcs10)
+
+        // signature
+        val pKey = caKey()
+        val sigGen = JcaContentSignerBuilder("SHA384WITHECDSA").build(pKey)
+        val holder: X509CertificateHolder = certificateBuilder.build(sigGen)
+        val x509CertificateStructure = holder.toASN1Structure()
+
+        // Read Certificate into x509 structure
+        val cf = CertificateFactory.getInstance("X.509", "BC")
+        val is1: InputStream = ByteArrayInputStream(x509CertificateStructure.encoded)
+        val theCert = cf.generateCertificate(is1) as X509Certificate
+        is1.close()
+
+        certificateRepository.save(Certificate.fromX509Certificate(theCert, caCertificate().subjectDN.name, chain()))
+        nutsCertificateRequestRepository.delete(request)
```

Comment:
This is in line with the current logic applied to Corda CSR's. CSR's are currently handed out-of-band in which a backup can be made. I'll add an issue for now to mark them as deleted in the DB rather than deleting them.